core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -507,86 +518,126 @@ trait EntityComponent {
+        val batchActions = createBatches(entityRefs.toSet).map(batch => lookupEntitiesByNames(workspaceId, batch))
+        DBIO.sequence(batchActions).map(_.flatten.toSeq).map { recs =>
+          recs.map(rec => AttributeEntityReference(rec.entityType, rec.name))
         }
       }
-    }
 
-    def getCopyConflicts(destWorkspaceContext: SlickWorkspaceContext, entitiesToCopy: TraversableOnce[Entity]): ReadAction[TraversableOnce[Entity]] = {
-      val entityQueries = entitiesToCopy.map { entity =>
-        findEntityByName(destWorkspaceContext.workspaceId, entity.entityType, entity.name).result.map {
-          case Seq() => None
-          case _ => Option(entity)
+      def getSoftConflicts(paths: Seq[Seq[AttributeEntityReference]]) = {
+        getCopyConflicts(destWorkspaceContext, paths.map(_.last)).map { conflicts =>
+          val conflictsAsRefs = conflicts.toSeq.map(_.toReference)
+          paths.filter(p => conflictsAsRefs.contains(p.last))
         }
       }
-      DBIO.sequence(entityQueries).map(_.toStream.collect { case Some(e) => e })
+
+      def buildSoftConflictTree(pathsRemaining: Seq[AttributeEntityReference], result: Seq[EntitySoftConflict]): Seq[EntitySoftConflict] = {
+        if(pathsRemaining.isEmpty) result
+        else Seq(EntitySoftConflict(pathsRemaining.head.entityType, pathsRemaining.head.entityName, buildSoftConflictTree(pathsRemaining.tail, result)))
+      }
+
+      val entitiesToCopyRefs = entityNames.map(name => AttributeEntityReference(entityType, name))
+
+      getHardConflicts(destWorkspaceContext.workspaceId, entitiesToCopyRefs).flatMap {
+        case Seq() => {
+          val pathsAndConflicts = for {
+            entityPaths <- getEntitySubtrees(sourceWorkspaceContext, entityType, entityNames)
+            softConflicts <- getSoftConflicts(entityPaths)
+          } yield (entityPaths, softConflicts)
+
+          pathsAndConflicts.flatMap { case (entityPaths, softConflicts) =>
+            if(softConflicts.isEmpty || linkExistingEntities) {
+              val allEntityRefs = entityPaths.flatten
+              val allConflictRefs = softConflicts.flatten
+              val entitiesToClone = allEntityRefs diff allConflictRefs
+
+              for {
+                entities <- list(sourceWorkspaceContext, entitiesToClone)
+                _ <- cloneEntities(destWorkspaceContext, entities.toSet, allConflictRefs)
+              } yield EntityCopyResponse(entities.map(_.toReference).toSeq, Seq.empty, Seq.empty)
+            } else {
+              val unmergedSoftConflicts = softConflicts.flatMap(c => buildSoftConflictTree(c, Seq.empty))
+              val grouped = unmergedSoftConflicts.groupBy(c => (c.entityType, c.entityName))
+              DBIO.successful(EntityCopyResponse(Seq.empty, Seq.empty, grouped.map{ case ((entityType, entityName), conflicts) => EntitySoftConflict(entityType, entityName, conflicts.flatMap(_.conflicts))}.toSeq))
```

Comment:
Thinking back on it, the method name `buildSoftConflictTree` is kind of a lie. It's more like `buildSoftConflictPaths`. But to illustrate what's going on, I'll build off of the same example from my other comment:

given the entity tree: 
```
sample_set
   ^- sampleA
      ^- participant1
   ^- sampleB
      ^- participant2
```

we should have something like the following for the paths:

```
Seq(
Seq(sample_set), 
Seq(sample_set, sampleA), 
Seq(sample_set, sampleA, participant1), 
Seq(sample_set, sampleB), 
Seq(sample_set, sampleB, participant2)
)
```

and `softConflicts` would end up being:

```
Seq(
Seq(sample_set, sampleA, participant1), 
Seq(sample_set, sampleB, participant2)
)
```

Assume `participant1` and `participant2` are soft conflicts. We need to merge those paths into a tree based on their root, otherwise we'll end up with something like:

```
Seq(
EntitySoftConflict(sample_set, sample_set, Seq(EntitySoftConflict(sample, sampleA, EntitySoftConflict(participant, participant1, Seq.empty))))
EntitySoftConflict(sample_set, sample_set, Seq(EntitySoftConflict(sample, sampleB, EntitySoftConflict(participant, participant2, Seq.empty))))
)
```

when really we want something like this:

```
Seq(
EntitySoftConflict(sample_set, sample_set, Seq(EntitySoftConflict(sample, sampleA, EntitySoftConflict(participant, participant1, Seq.empty)), EntitySoftConflict(sample, sampleB, EntitySoftConflict(participant, participant2, Seq.empty))))
)
```

I will rename some variables/functions related to building the conflict tree as well as leave some comments.